### PR TITLE
use `default/kubernetes` to test for dns

### LINKED ISSUE #464: Connectivity test doesn't support cluster name parameterization
Looks like the K8s cluster name is hardcoded to _kube-dns.kube-system.svc.cluster.local_ on line 511 of connectivity/check/deployment.go. I'm not running Kube-DNS, only CoreDNS, and my cluster name is customized. This results in failure of the initial kube-dns lookup since the address _kube-dns.kube-system.svc.cluster.local_ doesn't actually exist.

```
ansibleuser@st1k8scp01a ~ $ cilium connectivity test --all-flows -v --print-flows -p
ℹ️  Monitor aggregation detected, will skip some flow validation steps
⌛ [st1k8s01] Waiting for deployments [client client2 echo-same-node] to become ready...
⌛ [st1k8s01] Waiting for deployments [echo-other-node] to become ready...
⌛ [st1k8s01] Waiting for CiliumEndpoint for pod cilium-test/client-7b7bf54b85-fl2pg to appear...
⌛ [st1k8s01] Waiting for CiliumEndpoint for pod cilium-test/client2-666976c95b-4lxd6 to appear...
⌛ [st1k8s01] Waiting for CiliumEndpoint for pod cilium-test/echo-other-node-f4d46f75b-qbkcl to appear...
⌛ [st1k8s01] Waiting for CiliumEndpoint for pod cilium-test/echo-same-node-745bd5c77-cmmxj to appear...
⌛ [st1k8s01] Waiting for Service cilium-test/echo-other-node to become ready...
⌛ [st1k8s01] Waiting for Service cilium-test/echo-same-node to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.32:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.32:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.30:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.30:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.26:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.26:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.31:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.31:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.29:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.29:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.34:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.34:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.33:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.33:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.35:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.35:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.27:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.27:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.28:30852 (cilium-test/echo-other-node) to become ready...
⌛ [st1k8s01] Waiting for NodePort 10.244.0.28:31541 (cilium-test/echo-same-node) to become ready...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-2msmf to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-7xzf6 to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-cc5pn to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-mjl4j to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-26nsj to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-9gpwn to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-dcz45 to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-pgrw6 to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-q729v to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for Cilium pod kube-system/cilium-w6w8x to have all the pod IPs in eBPF ipcache...
⌛ [st1k8s01] Waiting for pod cilium-test/client-7b7bf54b85-fl2pg to reach kube-dns service...

Error: Connectivity test failed: timeout reached waiting lookup for kube-dns.kube-system.svc.cluster.local from pod cilium-test/client-7b7bf54b85-fl2pg to succeed

ansibleuser@st1k8scp01a ~ $ kubectl -n cilium-test exec -it client-7b7bf54b85-fl2pg -- nslookup coredns.kube-system.svc.cluster.local
Server:         192.168.0.3
Address:        192.168.0.3:53

** server can't find coredns.kube-system.svc.cluster.local: NXDOMAIN

** server can't find coredns.kube-system.svc.cluster.local: NXDOMAIN

command terminated with exit code 1
ansibleuser@st1k8scp01a ~ $ kubectl -n cilium-test exec -it client-7b7bf54b85-fl2pg -- nslookup coredns.kube-system.svc.st1k8s01     
Server:         192.168.0.3
Address:        192.168.0.3:53

Name:   coredns.kube-system.svc.st1k8s01
Address: 192.168.0.3
```

Cilium version in case someone wants it

```
ansibleuser@st1k8scp01a ~ $ cilium version                        
cilium-cli: v0.8.5 compiled with go1.16.5 on linux/amd64
cilium image (default): v1.10.2
cilium image (stable): v1.10.3
```

### PR DESCRIPTION
**Pull request by zhaojizhuang**
fix #464 
use `default/kubernetes` to test for dns

Signed-off-by: zhaojizhuang <571130360@qq.com>

### CODE REVIEW

**Comment by gandro** [comment]
The test failures look legitimate:
```
server can't find kubernetes.default: NXDOMAIN
```

**Comment by zhaojizhuang** [comment]
cc @gandro

**Review by tklauser — CHANGES_REQUESTED** [review]
@zhaojizhuang please squash the commits of this PR into one and add a commit message describing the intent of the change.

**Comment by zhaojizhuang** [comment]
> @zhaojizhuang please squash the commits of this PR into one and add a commit message describing the intent of the change.

@tklauser Done

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks for the revision. I left an inline comment why this new approach however will not work. Overall, the previous approach of using `nslookup kubernetes.default` was the correct one, as it will rely on `nslookup` doing the /etc/resolv.conf parsing for us.

The problem seems to be that our client docker image ships with `busybox` nslookup, which is known to be broken: https://github.com/kubernetes/kubernetes/issues/66924

I did a quick test locally, and if I install the upstream `nslookup` from bind-tools in the container, then `nslookup kubernetes.default` succeeds. My proposal therefore is to do the following:

1. Install the proper (non-busybox) version of `nslookup` from bind-tools in the images used for the connectivity test. We can add `bind-tools` here:
https://github.com/cilium/alpine-curl/blob/bcb227dc6431447fce30bfa8f80ebb1442729eb3/Dockerfile#L13
2. Release a new version of the `alpine-curl` image (I can help you with that) such that it contains a correct version of `nslookup`
3. Revert this to the previous approach of just invoking `nslookup kubernetes.default`, with the proper `nslookup` version, it will then actually do the right thing and parse the `/etc/resolv.conf` for you.

What do you think?

File: connectivity/check/util.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+// GetServiceHostname returns the fully qualified service hostname
+func GetServiceHostname(name, namespace string) string {
+	return fmt.Sprintf("%s.%s.svc.%s", name, namespace, GetClusterDomainName())
+}
+
+// GetClusterDomainName returns cluster's domain name
+func GetClusterDomainName() string {
+	once.Do(func() {
+		f, err := os.Open(resolverFileName)
```
[the remark this answers is not in the document]
Comment (zhaojizhuang) [reply connectivity/check/util.go:44]: @gandro For Pod running in the k8s cluster, it's `/etc/resolv.conf` if generated according to the `{{ .DNSDomain }}` https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/dns/manifests.go#L162 

Then we get the clusterdomain by this function `GetClusterDomainName `

for cluster 
```bash
    .:53 {
        errors
        health
        ready
        kubernetes example-cluster in-addr.arpa ip6.arpa {
          pods insecure
          upstream /etc/resolv.conf
          fallthrough in-addr.arpa ip6.arpa
        }
```

and the `/etc/resolv.conf` in pod is 
```bash
/ # cat /etc/resolv.conf
nameserver 169.254.25.10
search testnamespace.svc.example-cluster. svc.example-cluster. example-cluster.
options ndots:5
```
certainly we can fix the alpine image

[the remark this answers is not in the document]
Comment (zhaojizhuang) [reply connectivity/check/util.go:44]: @gandro https://github.com/cilium/alpine-curl/pull/14

**Review by tklauser — CHANGES_REQUESTED** [review]
A a new version v1.4.0 of the `alpine-curl` image was released with the fix. So will also need to update to that new version here once the corresponding image is built and pushed to quay.io:

https://github.com/cilium/cilium-cli/blob/254a814c6603c3ea22d63c14b83be69c4735f098/defaults/defaults.go#L75

**Review by gandro — APPROVED** [review]

**Review by tklauser — APPROVED** [review]

**Pull request merged**